.github/workflows/gateway-conformance.yml

```
@@ -0,0 +1,72 @@
+name: Gateway Conformance
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+
+jobs:
+  gateway-conformance:
+    runs-on: ubuntu-latest
+    steps:
+      # 1. Start the Kubo gateway
+      - name: Download Kubo gateway
+        uses: ipfs/download-ipfs-distribution-action@v1
+      - name: Start Kubo gateway
+        uses: ipfs/start-ipfs-daemon-action@v1
+
+      # 2. Download the gateway-conformance fixtures
+      - name: Download gateway-conformance fixtures
+        # TODO: use a release tag once https://github.com/ipfs/gateway-conformance/pull/19 is merged
+        uses: ipfs/gateway-conformance/.github/actions/extract-fixtures@e320c2b1f576f69732d824c4d48a1c37bb0ad320
+        with:
+          output: fixtures
+
+      # 3. Populate the Kubo gateway with the gateway-conformance fixtures
+      - name: Import fixtures
+        run: find ./fixtures -name '*.car' -exec ipfs dag import --pin-roots=false {} \;
+
+      # 4. Build the bifrost-gateway
+      - name: Setup Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.19.x
+      - name: Checkout bifrost-gateway
+        uses: actions/checkout@v3
+        with:
+          path: bifrost-gateway
+      - name: Build bifrost-gateway
+        run: go build
+        working-directory: bifrost-gateway
+
+      # 5. Start the bifrost-gateway
+      - name: Start bifrost-gateway
+        env:
+          PROXY_GATEWAY_URL: http://127.0.0.1:8080
+          GATEWAY_CONFORMANCE_TEST: true
+        run: ./bifrost-gateway &
+        working-directory: bifrost-gateway
+
+      # 6. Run the gateway-conformance tests
+      - name: Run gateway-conformance tests
+        # TODO: use a release tag once https://github.com/ipfs/gateway-conformance/pull/19 is merged
+        uses: ipfs/gateway-conformance/.github/actions/test@e320c2b1f576f69732d824c4d48a1c37bb0ad320
+        with:
+          gateway-url: http://127.0.0.1:8081
+          json: output.json
+          xml: output.xml
+          html: output.html
+          markdown: output.md
+          args: -skip TestGatewayCar
```

I think the fix for [the failing test](https://github.com/ipfs/bifrost-gateway/actions/runs/4511433301#summary-12252680845) is in https://github.com/ipfs/go-libipfs/pull/177  which is waiting for upstream fix in golang.

@galargh @laurentsenta  Is it possible to move `Content-Length` check into separate test that we can ignore for now, instead of entire  `TestGatewayCar`?
Or maybe just comment it out for now in gateway-compliance with `TODO` for later?

---

We can be more specific about the exclusion:
```suggestion
          args: -skip 'TestGatewayCar/GET_response_for_application/vnd.ipld.car'
```

---

Ah, but that's the only case we have there, so it doesn't change anything. Not ideal, but I think doing a separate case just for content length would be the most correct-ish thing to do here.

---

Wouldn't the fix for the underlying issue be a pretty simple one on the boxo side though? https://github.com/ipfs/boxo/pull/177#issuecomment-1484666302

---

I figured out how to make it possible to exclude a single check in a general way - https://github.com/ipfs/gateway-conformance/pull/23

Once the PR is merged, we'll be able to modify the exclusion here to `-skip 'TestGatewayCar/GET_response_for_application/vnd.ipld.car/Header_Content-Length'`.

---

@galargh it is, but it depends on Go itself: https://github.com/golang/go/pull/58578

---

@galargh I don't understand what you mean, but I think the problem is that Go sets `Content-Length`, even if you `nil` it.

---

Yeah, that's what I got too. That when you make a request, the response always contains `Content-Length`. Even if you explicitly ask it not to. What I'm saying is, can we force-nil the `Content-Length` header in the response after we make the request?

---

@galargh I'm confused. I thought this was a gateway test where we perform a GET to the gateway and check the headers. When Kubo replies, Go sets the header. The response will be in the test. What good would it make to nil the header in the test? Then we would not be really testing if the header is present or not. This to say that I don't understand what your idea is, or where the request and response are in Kubo.

---

I talked with Laurent. I was completely misunderstanding where the response header writing happens and where the problem is. Sorry for the confusion. 

I'm going to resolve this conversation now since we're only excluding the content-length header check now.